Họ tên: Huỳnh Thanh Sang

MSSV: 1712718

---

## Import

In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

## Lấy dữ liệu

Dữ liệu sử dụng trong đồ án này là file "beer_reviews.csv" đây là dữ liệu đánh giá về bia từ BeerAdvocates. Đây là dataset được thảo luận trong buổi nói chuyện "How to hire and test for data skills: A one-size-fits-all interview kit" từ https://conferences.oreilly.com/strata/strata-ny-2017/public/schedule/detail/59542. Dữ liệu lớn nên không thể upload lên github nên mình sẽ để link [ở đây](https://www.kaggle.com/rdoume/beerreviews)


## Khám phá dữ liệu

In [2]:
beer_reviews_df = pd.read_csv("beer_reviews.csv")
beer_reviews_df

brewery_id                 brewery_name  review_time  review_overall  \
0             10325              Vecchio Birraio   1234817823             1.5   
1             10325              Vecchio Birraio   1235915097             3.0   
2             10325              Vecchio Birraio   1235916604             3.0   
3             10325              Vecchio Birraio   1234725145             3.0   
4              1075      Caldera Brewing Company   1293735206             4.0   
...             ...                          ...          ...             ...   
1586609       14359  The Defiant Brewing Company   1162684892             5.0   
1586610       14359  The Defiant Brewing Company   1161048566             4.0   
1586611       14359  The Defiant Brewing Company   1160702513             4.5   
1586612       14359  The Defiant Brewing Company   1160023044             4.0   
1586613       14359  The Defiant Brewing Company   1160005319             5.0   

         review_aroma  review_appearance review_profilename  \
0                 2.0                2.5            stcules   
1                 2.5                3.0            stcules   
2                 2.5                3.0            stcules   
3                 3.0                3.5            stcules   
4                 4.5                4.0     johnmichaelsen   
...               ...                ...                ...   
1586609           4.0                3.5         maddogruss   
1586610           5.0                2.5          yelterdow   
1586611           3.5                3.0           TongoRad   
1586612           4.5                4.5           dherling   
1586613           4.5                4.5               cbl2   

                             beer_style  review_palate  review_taste  \
0                            Hefeweizen            1.5           1.5   
1                    English Strong Ale            3.0           3.0   
2                Foreign / Export Stout            3.0           3.0   
3                       German Pilsener            2.5           3.0   
4        American Double / Imperial IPA            4.0           4.5   
...                                 ...            ...           ...   
1586609                     Pumpkin Ale            4.0           4.0   
1586610                     Pumpkin Ale            2.0           4.0   
1586611                     Pumpkin Ale            3.5           4.0   
1586612                     Pumpkin Ale            4.5           4.5   
1586613                     Pumpkin Ale            4.5           4.5   

                      beer_name  beer_abv  beer_beerid  
0                  Sausa Weizen       5.0        47986  
1                      Red Moon       6.2        48213  
2        Black Horse Black Beer       6.5        48215  
3                    Sausa Pils       5.0        47969  
4                 Cauldron DIPA       7.7        64883  
...                         ...       ...          ...  
1586609      The Horseman's Ale       5.2        33061  
1586610      The Horseman's Ale       5.2        33061  
1586611      The Horseman's Ale       5.2        33061  
1586612      The Horseman's Ale       5.2        33061  
1586613      The Horseman's Ale       5.2        33061  

[1586614 rows x 13 columns]

Dữ liệu có 1586614 dòng và 13 cột

## Ý nghĩa của mỗi dòng và cột

+ Mỗi dòng là một đánh giá của từng người về một loại bia nhất định nào đó.
+ Ý nghĩa mỗi cột:
    - **brewery_id**: Số ID của nhà máy sản xuất bia.
    - **brewery_name**: Tên nhà máy ứng với id của nhà máy.
    - **review_time**: Ngày review.
    - **review_overall**: Điểm review tổng thể của bia.
    - **review_aroma**: Điểm review về mùi thơm của bia.
    - **review_appearance**: Điểm review về bề ngoài của chai bia.
    - **review_profilename**: Tên profile của người review.
    - **beer_style**: Phong cách của bia :)).
    - **review_palate**: Điểm cho cảm giác của bia để lại trên vòm miệng.
    - **review_taste**: Điểm cho vị của bia.
    - **beer_name**: Tên của bia.
    - **beer_abv**: Nồng độ cồn của bia.
    - **beer_beerid**: Số ID của bia

## Kiểu dữ liệu và số lượng giá trị khác rỗng mỗi cột

In [3]:
beer_reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1586614 entries, 0 to 1586613
Data columns (total 13 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   brewery_id          1586614 non-null  int64  
 1   brewery_name        1586599 non-null  object 
 2   review_time         1586614 non-null  int64  
 3   review_overall      1586614 non-null  float64
 4   review_aroma        1586614 non-null  float64
 5   review_appearance   1586614 non-null  float64
 6   review_profilename  1586266 non-null  object 
 7   beer_style          1586614 non-null  object 
 8   review_palate       1586614 non-null  float64
 9   review_taste        1586614 non-null  float64
 10  beer_name           1586614 non-null  object 
 11  beer_abv            1518829 non-null  float64
 12  beer_beerid         1586614 non-null  int64  
dtypes: float64(6), int64(3), object(4)
memory usage: 157.4+ MB


Các cột dạng categorical là: brewery_name, review_profilename, beer_style, beer_name.Các cột còn là là brewery_id, review_time, review_overall, review_aroma, review_appearance, review_palate, review_taste, beer_abv, beer_beerid.

## Số lượng giá trị thiếu mỗi cột

In [4]:
beer_reviews_df.isnull().sum()

brewery_id                0
brewery_name             15
review_time               0
review_overall            0
review_aroma              0
review_appearance         0
review_profilename      348
beer_style                0
review_palate             0
review_taste              0
beer_name                 0
beer_abv              67785
beer_beerid               0
dtype: int64

## Tiền xử lí dữ liệu

**Chuyển đổi kiểu giá trị cột review_time về dạng datetime**

Ta thấy cột review_time đang có kiểu dữ liệu là int64 và chúng ta sẽ không biết được ngày mà reviewer làm bài đánh giá

In [5]:
beer_reviews_df.review_time = pd.to_datetime (beer_reviews_df.review_time,unit = 's')

**Xóa các dòng bị thiếu giá trị**

In [6]:
beer_reviews_df=beer_reviews_df.dropna()

**Xóa các dòng bị trùng**

Ở đây chúng ta phải xóa các dòng bị trùng vì có thể người đánh giá đã đánh giá nhiều lần để giữ thứ hạng cho dòng bia mà mình yêu thích.

In [7]:
beer_reviews_df = beer_reviews_df.sort_values("review_overall",ascending=False)
beer_reviews_df = beer_reviews_df.drop_duplicates(subset = ['review_profilename','beer_beerid'],keep='first')
beer_reviews_df = beer_reviews_df.drop_duplicates(subset = ['review_profilename','beer_name'],keep='first')

In [8]:
beer_reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1496263 entries, 1586613 to 39648
Data columns (total 13 columns):
 #   Column              Non-Null Count    Dtype         
---  ------              --------------    -----         
 0   brewery_id          1496263 non-null  int64         
 1   brewery_name        1496263 non-null  object        
 2   review_time         1496263 non-null  datetime64[ns]
 3   review_overall      1496263 non-null  float64       
 4   review_aroma        1496263 non-null  float64       
 5   review_appearance   1496263 non-null  float64       
 6   review_profilename  1496263 non-null  object        
 7   beer_style          1496263 non-null  object        
 8   review_palate       1496263 non-null  float64       
 9   review_taste        1496263 non-null  float64       
 10  beer_name           1496263 non-null  object        
 11  beer_abv            1496263 non-null  float64       
 12  beer_beerid         1496263 non-null  int64         
dtypes: datet

**Loại bỏ điểm outlier**

In [9]:
round(beer_reviews_df.describe(),2) #Làm tròn 2 chữ số thập phân 

brewery_id  review_overall  review_aroma  review_appearance  \
count  1496263.00      1496263.00    1496263.00         1496263.00   
mean      3068.78            3.83          3.75               3.85   
std       5541.95            0.72          0.70               0.61   
min          1.00            0.00          1.00               0.00   
25%        141.00            3.50          3.50               3.50   
50%        413.00            4.00          4.00               4.00   
75%       2250.00            4.50          4.00               4.00   
max      28003.00            5.00          5.00               5.00   

       review_palate  review_taste    beer_abv  beer_beerid  
count     1496263.00    1496263.00  1496263.00   1496263.00  
mean            3.75          3.80        7.05     21445.02  
std             0.68          0.73        2.33     21800.60  
min             1.00          1.00        0.01         5.00  
25%             3.50          3.50        5.20      1655.00  
50%             4.00          4.00        6.50     12962.00  
75%             4.00          4.50        8.50     39286.00  
max             5.00          5.00       57.70     77316.00

Ta thấy trong phần đánh giá điểm thì thang điểm sẽ được giới hạn trong [1:5] nhưng ta thấy trong cột review_overall và cột review_appearance có giá trị min là 0.00 nên ta cần loại bỏ những điểm outlier này.

In [10]:
beer_reviews_df = beer_reviews_df.loc[(beer_reviews_df.review_overall>=1) & (beer_reviews_df.review_appearance>=1)]

In [11]:
round(beer_reviews_df.describe(),2)

brewery_id  review_overall  review_aroma  review_appearance  \
count  1496256.00      1496256.00    1496256.00         1496256.00   
mean      3068.78            3.83          3.75               3.85   
std       5541.96            0.72          0.70               0.61   
min          1.00            1.00          1.00               1.00   
25%        141.00            3.50          3.50               3.50   
50%        413.00            4.00          4.00               4.00   
75%       2250.00            4.50          4.00               4.00   
max      28003.00            5.00          5.00               5.00   

       review_palate  review_taste    beer_abv  beer_beerid  
count     1496256.00    1496256.00  1496256.00   1496256.00  
mean            3.75          3.80        7.05     21445.10  
std             0.68          0.73        2.33     21800.62  
min             1.00          1.00        0.01         5.00  
25%             3.50          3.50        5.20      1655.00  
50%             4.00          4.00        6.50     12962.00  
75%             4.00          4.50        8.50     39286.00  
max             5.00          5.00       57.70     77316.00

**Sau khi xóa các điểm outlier thì dữ liệu còn lại các giá trị nằm trong đoạn từ [1:5]**

In [12]:
num_cols = beer_reviews_df.shape[1]
num_rows = beer_reviews_df.shape[0]

print("Kết thúc giai đoạn tiền xử lí")
print("Số dòng: ", num_rows) 
print("Số cột: ", num_cols)

Kết thúc giai đoạn tiền xử lí
Số dòng:  1496256
Số cột:  13


In [13]:
beer_reviews_df.head(5)

brewery_id                 brewery_name         review_time  \
1586613       14359  The Defiant Brewing Company 2006-10-04 23:41:59   
1054114         718              Brouwerij Bavik 2006-08-03 11:48:03   
783659        13014        Surly Brewing Company 2011-06-13 13:00:36   
783658        13014        Surly Brewing Company 2011-06-27 15:25:22   
1182432       10099         Dogfish Head Brewery 2005-12-01 04:09:23   

         review_overall  review_aroma  review_appearance review_profilename  \
1586613             5.0           4.5                4.5               cbl2   
1054114             5.0           4.0                4.5        BEERchitect   
783659              5.0           5.0                5.0           deadbody   
783658              5.0           5.0                4.0              pdeme   
1182432             5.0           4.5                4.0             Sayson   

                               beer_style  review_palate  review_taste  \
1586613                       Pumpkin Ale            4.5           4.5   
1054114                           Witbier            4.5           4.0   
783659                 American Brown Ale            5.0           5.0   
783658                 American Brown Ale            4.5           5.0   
1182432  American Double / Imperial Stout            4.5           4.0   

                  beer_name  beer_abv  beer_beerid  
1586613  The Horseman's Ale      5.20        33061  
1054114          Wittekerke      5.00         2477  
783659         Moe's Bender      5.52        58120  
783658         Moe's Bender      5.52        58120  
1182432    World Wide Stout     18.00         2392

## Đưa ra câu hỏi